lib/src/migrator.dart

```diff
@@ -0,0 +1,247 @@
+// Copyright 2018 Google LLC
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+// The sass package's API is not necessarily stable. It is being imported with
+// the Sass team's explicit knowledge and approval. See
+// https://github.com/sass/dart-sass/issues/236.
+import 'package:sass/src/ast/sass.dart';
+import 'package:sass/src/visitor/recursive_statement.dart';
+import 'package:sass/src/visitor/interface/expression.dart';
+
+import 'package:path/path.dart' as p;
+
+import 'built_in_functions.dart';
+import 'local_scope.dart';
+import 'patch.dart';
+import 'stylesheet_migration.dart';
+import 'utils.dart';
+
+/// Runs a migration of [entrypoints] and their dependencies without writing
+/// any changes to disk.
+p.PathMap<String> migrateFiles(List<String> entrypoints) =>
+    _Migrator().migrate(entrypoints);
+
+class _Migrator extends RecursiveStatementVisitor implements ExpressionVisitor {
+  /// List of all migrations for files touched by this run.
+  p.PathMap<StylesheetMigration> _migrations = p.PathMap();
+
+  /// List of migrations in progress. The last item in the current migration.
+  List<StylesheetMigration> _activeMigrations = [];
+
+  StylesheetMigration get _currentMigration =>
+      _activeMigrations.isNotEmpty ? _activeMigrations.last : null;
+
+  /// Runs the migrator on each item in [entrypoints] in turn and returns a
+  /// map of migrated contents.
+  p.PathMap<String> migrate(List<String> entrypoints) {
+    for (var entrypoint in entrypoints) {
+      _migrateStylesheet(entrypoint);
+    }
+    var results = p.PathMap<String>();
+    for (var migration in _migrations.values) {
+      results[migration.path] = migration.migratedContents;
+    }
+    return results;
+  }
+
+  /// Migrates the stylesheet at [path] if it hasn't already been migrated and
+  /// returns the StylesheetMigration instance for it regardless.
+  StylesheetMigration _migrateStylesheet(String path) {
+    path = canonicalizePath(path,
+        loadPath: _currentMigration == null
+            ? p.current
+            : p.dirname(_currentMigration.path));
+    if (_migrations.containsKey(path)) return _migrations[path];
```

Comment:
I think you can clean this up a bit using [`Map.putIfAbsent()`](https://api.dartlang.org/stable/2.1.0/dart-core/Map/putIfAbsent.html).